### DISCUSSION TITLE
Cadquery Nptyping Error

### DISCUSSION THREAD

**Discussion by Xero-Cool** [joshreve/dactyl-keyboard#84] [Q&A] [answered]
I ran into the issue with dependency conflicts when using the autodocker install as mentioned here #77 and decided to go down the Conda route. The solid/OpenSCAD engine was working, but I was running into an nptyping error when trying to run dactyl_manuform.py with the cadquery/OpenCascade engine. 

Here is the output if you are interested.

> (dactyl-keyboard) C:\Users\jchil\dactyl-keyboard-44>python src/dactyl_manuform.py
> NO CONFIGURATION SPECIFIED, USING run_config.json
> Found Current Engine in Config = cadquery
> Traceback (most recent call last):
>   File "src/dactyl_manuform.py", line 70, in <module>
>     from helpers_cadquery import *
>   File "C:\Users\jchil\dactyl-keyboard-44\src\helpers_cadquery.py", line 1, in <module>
>     import cadquery as cq
>   File "C:\Users\jchil\anaconda3\envs\dactyl-keyboard\lib\site-packages\cadquery\__init__.py", line 31, in <module>
>     from .assembly import Assembly, Color, Constraint
>   File "C:\Users\jchil\anaconda3\envs\dactyl-keyboard\lib\site-packages\cadquery\assembly.py", line 10, in <module>
>     from .occ_impl.solver import (
>   File "C:\Users\jchil\anaconda3\envs\dactyl-keyboard\lib\site-packages\cadquery\occ_impl\solver.py", line 22, in <module>
>     class ConstraintSolver(object):
>   File "C:\Users\jchil\anaconda3\envs\dactyl-keyboard\lib\site-packages\cadquery\occ_impl\solver.py", line 89, in ConstraintSolver
>     Callable[[Array[(Any,), float]], float],
>   File "C:\Users\jchil\anaconda3\envs\dactyl-keyboard\lib\site-packages\nptyping\ndarray.py", line 73, in __getitem__
>     shape_expression, dtype = _get_from_tuple(item)
>   File "C:\Users\jchil\anaconda3\envs\dactyl-keyboard\lib\site-packages\nptyping\ndarray.py", line 129, in _get_from_tuple
>     f"Unexpected argument '{item[0]}', expecting"
> nptyping.error.InvalidArgumentsError: Unexpected argument '(typing.Any,)', expecting Shape[<ShapeExpression>] or Literal[<ShapeExpression>]

After some searching, it looks like a new release of np-typing (2.0) that was recently pushed conflicts with cadquery. A workaround seems to be to perform the cadquery installation with an older version of nptyping. [Here is the relevant link.](https://github.com/CadQuery/cadquery/issues/1046) 

`conda install -c cadquery -c conda-forge cadquery=master nptyping=1.4.4 `

Looks like a fix has been merged with the master in cadquery, but it's not released yet. I think... I don't really know how this stuff works, I just bang my hands on the keyboard until the screen does something I like.

**Comment by chpf** [comment, marked as answer]
This is what breaks the docker build as well. I simply added `nptyping=1.4.4` to `environment.yml`.

**Reply by joshreve** [reply]
Cadquery makes such good geometry, but it is insanely fragile.  I'll add this to the environment file.  That is my normal approach for the eternal battle to keep cadquery running.

**Reply by FredBelleville** [reply]
Hi guys, sorry about noob question here. I can't see the correction nptyping=1.4.4 in the environment.yml. In which branch is it suppose to have been commited to?

I also have no clue on how to run the conda install command line? Could someone please help me? Thanks!

**Reply by chpf** [reply]
You can just add it to the file yourself. When using docker (which is what the evironment.yml file is for) you don't have to bother with using conda.

**Reply by FredBelleville** [reply]
I try something like this but still the same get error. I am guessing I don't have the right syntax?

name: myenv
channels:
  - cadquery
  - conda-forge
  - defaults
dependencies:
  - python>=3.6
  - cadquery=2
  - pip
  - numpy
  - scipy
  - pip:
    - solidpython
    - nptyping=1.4.4

**Reply by joshreve** [reply]
I unfortunately have little time a the moment, so I can't test the solution, but try adding the nptyping line above `pip:` in the main section.  The lower pip section is only for libraries only available through pip.  The above section will use conda for the mini-conda environment.  If that doesn't work let me know and I can try to take a look, but I'm not sure when that will be.
